ExecutionLifecycleTest.when_processorProcessThrows_then_failJob

The `ExecutionLifecycleTest.when_processorProcessThrows_then_failJob` failed in PR builder on `master` branch:
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/5119/

Stacktrace:
```
java.lang.AssertionError: expected:<4> but was:<6>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at com.hazelcast.jet.core.ExecutionLifecycleTest.assertPClosedWithError(ExecutionLifecycleTest.java:765)
	at com.hazelcast.jet.core.ExecutionLifecycleTest.when_processorProcessThrows_then_failJob(ExecutionLifecycleTest.java:327)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/hazelcast/src/main/java/com/hazelcast/jet/impl/JobExecutionService.java
+++ b/hazelcast/src/main/java/com/hazelcast/jet/impl/JobExecutionService.java
@@ -491,7 +491,7 @@ public class JobExecutionService implements DynamicMetricsProvider {
                     }
                     return terminalMetrics;
                 })
-                .whenComplete(withTryCatch(logger, (i, e) -> {
+                .whenCompleteAsync(withTryCatch(logger, (i, e) -> {
                     completeExecution(execCtx, peel(e));
 
                     if (e instanceof CancellationException) {

--- a/hazelcast/src/main/java/com/hazelcast/jet/impl/execution/ExecutionContext.java
+++ b/hazelcast/src/main/java/com/hazelcast/jet/impl/execution/ExecutionContext.java
@@ -55,6 +55,7 @@ import java.util.Set;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Function;
 
 import static com.hazelcast.jet.Util.idToString;
@@ -113,6 +114,7 @@ public class ExecutionContext implements DynamicMetricsProvider {
     private volatile RawJobMetrics jobMetrics = RawJobMetrics.empty();
 
     private InternalSerializationService serializationService;
+    private final AtomicBoolean executionCompleted = new AtomicBoolean();
 
     public ExecutionContext(NodeEngine nodeEngine, long jobId, long executionId, boolean isLightJob) {
         this.jobId = jobId;
@@ -199,6 +201,7 @@ public class ExecutionContext implements DynamicMetricsProvider {
         synchronized (executionLock) {
             if (executionFuture != null) {
                 // beginExecution was already called or execution was cancelled before it started.
+                LoggingUtil.logFine(logger, "%s: execution started after cancelled", jobNameAndExecutionId());
                 return executionFuture;
             } else {
                 // begin job execution
@@ -229,6 +232,10 @@ public class ExecutionContext implements DynamicMetricsProvider {
     public void completeExecution(Throwable error) {
         assert executionFuture == null || executionFuture.isDone()
                 : "If execution was begun, then completeExecution() should not be called before execution is done.";
+
+        if (!executionCompleted.compareAndSet(false, true)) {
+            return;
+        }
 
         for (Tasklet tasklet : tasklets) {
             try {

--- a/hazelcast/src/test/java/com/hazelcast/jet/core/ExecutionLifecycleTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/jet/core/ExecutionLifecycleTest.java
@@ -51,7 +51,6 @@ import com.hazelcast.test.annotation.ParallelJVMTest;
 import com.hazelcast.test.annotation.QuickTest;
 import org.junit.Before;
 import org.junit.BeforeClass;
-import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
@@ -103,7 +102,6 @@ import static org.junit.Assume.assumeTrue;
 @RunWith(Parameterized.class)
 @UseParametersRunnerFactory(HazelcastSerialParametersRunnerFactory.class)
 @Category({QuickTest.class, ParallelJVMTest.class})
-@Ignore // https://github.com/hazelcast/hazelcast/issues/18707
 public class ExecutionLifecycleTest extends SimpleTestInClusterSupport {
 
     private static final int MEMBER_COUNT = 2;
@@ -747,8 +745,11 @@ public class ExecutionLifecycleTest extends SimpleTestInClusterSupport {
 
     private void assertPsClosedWithError() {
         assertEquals(MEMBER_COUNT, MockPS.initCount.get());
-        assertEquals(MEMBER_COUNT, MockPS.closeCount.get());
-        assertEquals(MEMBER_COUNT, MockPS.receivedCloseErrors.size());
+        // with light jobs the init can be called on not all the members - the execution on one member
+        // can be cancelled due to the failure on the other member before it was initialized.
+        int minCount = useLightJob ? 1 : MEMBER_COUNT;
+        assertBetween("close count", MockPS.closeCount.get(), minCount, MEMBER_COUNT);
+        assertBetween("received close errors", MockPS.receivedCloseErrors.size(), minCount, MEMBER_COUNT);
 
         for (int i = 0; i < MEMBER_COUNT; i++) {
             assertOneOfExceptionsInCauses(MockPS.receivedCloseErrors.get(i),
@@ -764,7 +765,10 @@ public class ExecutionLifecycleTest extends SimpleTestInClusterSupport {
     }
 
     private void assertPClosedWithError() {
-        assertEquals(MEMBER_COUNT * parallelism, MockP.closeCount.get());
+        // with light jobs the init can be called on not all the members - the execution on one member
+        // can be cancelled due to the failure on the other member before it was initialized.
+        int minCount = useLightJob ? parallelism : MEMBER_COUNT * parallelism;
+        assertBetween("close count", MockP.closeCount.get(), minCount, MEMBER_COUNT * parallelism);
     }
 
     private void assertOneOfExceptionsInCauses(Throwable caught, Throwable... expected) {

